src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileParser.Cache.cs

```diff
@@ -0,0 +1,86 @@
+﻿extern alias __uno;
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Uno.UI.SourceGenerators.XamlGenerator;
+
+internal partial class XamlFileParser
+{
+	private struct CachedFileKey : IEquatable<CachedFileKey>
+	{
+		public CachedFileKey(string includeXamlNamespaces, string excludeXamlNamespaces, string file, ImmutableArray<byte> checksum)
+		{
+			IncludeXamlNamespaces = includeXamlNamespaces;
+			ExcludeXamlNamespaces = excludeXamlNamespaces;
+			File = file;
+			Checksum = checksum;
+		}
+
+		public string IncludeXamlNamespaces { get; }
+		public string ExcludeXamlNamespaces { get; }
+		public string File { get; }
+		public ImmutableArray<byte> Checksum { get; }
+
+		public override bool Equals(object? obj) => obj is CachedFileKey key && Equals(key);
+
+		public bool Equals(CachedFileKey other) =>
+			IncludeXamlNamespaces == other.IncludeXamlNamespaces &&
+			ExcludeXamlNamespaces == other.ExcludeXamlNamespaces &&
+			File == other.File &&
+			Checksum.Equals(other.Checksum);
```

Comment:
TODO: This is a reference equality for the underlying array of `ImmutableArray`.
This might not be correct.